VM Support Refinement

jrajahalme opened the pull request:
Split IPAM bootstrap into configuration and startup phases so that the configuration can still be modified in between, and join cluster (if --join-cluster is set) before starting IPAM. This allows cluster to provide cluster CIDR config for the VM.

Add CiliumExternalWorkload CRD that dictates the namespace and labels for an external workload, such as a VM. VM's hostname and CEW CRD name must match, and the VM is not admitted if a matching CEW CRD is not found.

Move 'hubble.tls.auto.cronJob' values to 'certgen' and add a 'hubble.' prefix to hubble specific helm partials.

Add helm templates and values for `clustermesh-apiserver`.

Move `cilium/clustermesh-apiserver` to the main Cilium repo.

Add `PROVISION_EXTERNAL_WORKLOAD` option to compile and run Cilium agent in docker in the runtime CI VM. This mirrors the planned deployment model for Cilium Agent in external workload VMs.

TODO:
- [x] Update `cilium/certgen` version and SHA after required changes have been merged. Before this `.clustermesh.apiserver.tls.auto.method='cronJob'` will fail.
- [x] Enable external workloads in `install/kubernetes/experimental-install.yaml` (depends on the `certgen` update)
- [ ] CI test(s)

For later PRs:
- [x] validate k8s service access from VMs
- [ ] test with direct routing modes, remove the need for IPAM setup in this case
- [ ] check if tunneling can be done without any IPAM (i.e., w/o creating `cilium_host` device)
- [x] wire cluster DNS to the VMs so that VMs can use cluster-defined DNS names
- [ ] pass policy from cluster to the VMs
- [ ] enforce policy in VMs

Fixes: #13309

tklauser commented on daemon/cmd/ipam.go:
```diff
@@ -338,7 +338,6 @@ func (d *Daemon) bootstrapIPAM() {
 	// Then, we will calculate the IPv4 or IPv6 alloc prefix based on the IPv6
 	// or IPv4 alloc prefix, respectively, retrieved by k8s node annotations.
 	bootstrapStats.ipam.Start()
```
I think this should be removed (as well as the corresponding `bootstrapStats.ipam.End(true)` at line 363) since the functionality is moved to `(*Daemon).startIPAM()` by this PR.

jrajahalme replied:
Done.

tklauser commented on pkg/node/store/store.go:
```diff
@@ -101,34 +106,85 @@ type NodeRegistrar struct {
+	if err != nil {
+		registerStore.Release()
+		return nil, err
+	}
```
Nit:

```suggestion
	if err != nil {
		registerStore.Release()
		return nil, err
	}

	// Wait until an updated key is received from the kvstore
	select {
	case n = <-registerObserver.updates:
	case <-time.After(defaults.NodeInitTimeout / 10):
		registerStore.Release()
		return nil, fmt.Errorf("timed out waiting for node identity")
	}
```

jrajahalme replied:
Applied, thanks!

aanm commented on daemon/cmd/daemon.go:
```diff
@@ -598,6 +598,10 @@ func NewDaemon(ctx context.Context, epMgr *endpointmanager.EndpointManager, dp d
 		node.SetLabels(agentLabels)
+
+		// This can override node addressing config, so do this before starting IPAM
+		log.Infof("Calling JoinCluster(%s)", nodeTypes.GetName())
```
log.WithFields.. for the `nodeTypes.GetName()`

jrajahalme replied:
Changing this to debug as well.

aanm commented on pkg/node/types/node.go:
```diff
@@ -134,7 +134,6 @@ func (n *Node) ToCiliumNode() *ciliumv2.CiliumNode {
 	}
 }
 
-
```
Can be squashed with the previous commit.

jrajahalme replied:
I lost the context for this. But in case this is for the first two commits (where `pkg/node/types/node.go`) is changed, I'll keep them separate until we get clean CI runs so that triaging will be easier if the IPAM bootstrap changes cause problems.

jrajahalme replied:
Squashed the first two commits together.

aanm reviewed (approved):
Nothing blocking from my side

jrajahalme committed:
helm: Move 'hubble.tls.auto.cronJob' values to 'certgen'

Certgen image can be used for generating certificates for multiple
targets (not only hubble), so it makes sense to not nest it's values
under hubble.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
helm: Add 'hubble.' prefix to hubble specific helm partials.

The helm partials for generating hubble certs are hubble specific, indicate that in their names.

Delete the unused 'ca.gen-certs'.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
install: Allow cilium-certgen to create and update external workload secrets

Update ClusterRole to allow cilium-certgen to create and update
external workload secrets/configmaps.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

christarazi commented on examples/crds/ciliumexternalworkloads.yaml:
```diff
@@ -0,0 +1,76 @@
+
```
Nit: you can squash this commit with the previous commit that makes changes to the CRD. It makes sense to combine, in case someone `git blame`s on the generated file, they'll be lead to the "wrong" commit.

jrajahalme replied:
Will do. I kept them separate so that I had a clean view which files I'm working with and which are merely generated :-)

christarazi commented on pkg/k8s/apis/cilium.io/v2/cew_types.go:
```diff
@@ -0,0 +1,99 @@
+//  Copyright 2020 Authors of Cilium
```
You'll need to add two more things:

1) Make the agent wait for this new CRD, see commit e91f803c3
2) As Andre mentioned, you'll need to add the RBAC permissions for the agent to access this new resource, see https://github.com/cilium/cilium/blob/17c4f80ea32495ecd62c4f1a39455c45b2210f6c/install/kubernetes/cilium/templates/cilium-agent-clusterrole.yaml#L90

jrajahalme replied:
Cilium agents are not accessing CiliumExternalWorkload CRDs, so IMO agent RBAC permissions do not need to be amended. Only `clustermesh-apiserver` reads CEW CRDs :-)

jrajahalme replied:
@christarazi Also, I'm not sure if `clustermesh-apiserver` needs to wait for CRDs or not? It is watching on a number of CRDs, as well as creating and updating CRDs.

aanm replied:
@jrajahalme If that's the case, cilium agent does not need any RBAC permissions to access the new resource. However, `clustermesh-apiserver` needs those RBAC permissions.

>@christarazi Also, I'm not sure if `clustermesh-apiserver` needs to wait for CRDs or not? It is watching on a number of CRDs, as well as creating and updating CRDs.

It depends, if `clustermesh-apiserver` depends on those CRDs then it needs to be sure they are available, and if not, it needs to wait for them.

jrajahalme replied:
OK, I'll need to add a wait for the CRDs before starting any of the CRD watchers in `clustermesh-apiserver`.

christarazi reviewed (changes requested):
Mostly reviewed K8s side of things; looks good besides a few items to address.

jrajahalme committed:
test: Add option to start runtime Cilium from docker

Add PROVISION_EXTERNAL_WORKLOAD environment variable that if not
"false" will cause runtime VM compile and run Cilium in docker with
configuration suitable for external workload to join a Cilium cluster.
This is using the new 'test/provision/externalworkload_install.sh',
which builds and starts a Cilium docker image for a VM.

TLS certs for the VM must be extracted from the k8s cluster before
starting the VM.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
install/kubernetes: Enable external workloads in experimental-install.yaml

Enable support for external workloads in the experimental install
using "cronJob" auto TLS method.

"helm" method would include the generated keys in
experimental-install.yaml which would be a bad idea in the repo as
then all users of experimental-install.yaml would be using the same
key.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
k8s: Add CiliumExternalWorkload CRD

CiliumExternalWorkload CRD is used to specify which VMs can join the
cluster. Only clustermesh-apiserver is watching for these.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
clustermesh-apiserver: Add to cilium repo

Move clustermesh-apiserver to Cilium repo to ease development.

Add support scripts for mTLS config management to contrib/k8s.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
daemon: Split IPAM bootstrap, join cluster in between

Split IPAM bootstrap into two phases:
1. Configure IPAM (configureIPAM())
2. Begin using the IPAM configuration (startIPAM())

This allows overriding the IPAM configuration set in phase 1
before starting the use of IPAM configuration in phase 2.

Split the --join-cluster feature init into two phases:

1. Get config from clustermesh-apiserver by resetting node
registration to only contain the node name. clustermesh-apiserver
updates the key with configuration for this node derived from the
cluster state.

2. Update node registration after node IPAM is started with full node
information.

When node registration is started, the local cluster name may not be
known yet. To accomodate this only the node name is used to identify
the node registration.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

gandro reviewed (approved):
I only reviewed the changes regarding `certgen` and `hubble.tls.auto` helm changes. They look good to me!

jrajahalme committed:
k8s: Add CEW CRD to CRD waiter

Add CiliumExternalWorkload CRD to the CRD waiter. Cilium agent does
not use CiliumExternalWorkload CRD but this is needed for the
clustermesh-apiserver, and including it also with Cilium agent does
not hurt.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

christarazi commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1339,3 +1342,124 @@ enableCriticalPriorityClass: true
+      # For GKE LoadBalancer, use annotation cloud.google.com/load-balancer-type: "Internal"
+      # For EKS LoadBalancer, use annotation service.beta.kubernetes.io/aws-load-balancer-internal: 0.0.0.0/0
+      annotations:
+      - test-annotation: "Foo"
```
This is probably not supposed to be included?

jrajahalme replied:
Right, will remove!

jrajahalme replied:
Fixed!

christarazi reviewed (approved):
LGTM assuming no further changes to this PR (i.e. CRD watching for the clustermesh-apiserver). Just one minor comment to resolve, doesn't require a second review.

jrajahalme committed:
k8s: Refactor CRD waiter to pkg/k8s/synced

Minimize WaitForCRDsToRegister() dependencies by making it a
standalone function and by moving k8sResoucesSynced and k8sAPIGroups
from pkg/k8s/watchers to pkg/k8s/synced.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
clustermesh-apiserver: Add wait for CRDs before launching watchers

CRD watchers fail if the CRD has not (yet) been registered. Add a wait
for CRDs before starting k8s watchers so that we get rid of the race
with cilium operator that registers the CRDs.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
install/kubernetes: Add templates for clustermesh-apiserver

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

christarazi commented on pkg/k8s/synced/crd.go:
```diff
@@ -163,34 +158,34 @@ func (k *K8sWatcher) crdsInit(ctx context.Context, watcherClient apiextclientset
 }
 
-func addCRD(crds *crdState, obj interface{}) {
+func (s *crdState) addCRD(obj interface{}) {
```
I think we can rename these now to simply `add` and just drop the `CRD` part completely as it's redundant.

jrajahalme replied:
Done, changed "deleteCRD" to "remove" as well.

christarazi commented on pkg/k8s/synced/crd.go:
```diff
@@ -231,25 +226,35 @@ type crdState struct {
 
-func newCRDState(m map[string]bool) crdState {
+func newCRDState() crdState {
+	m := make(map[string]bool, len(crdResourceNames))
+	for _, name := range crdResourceNames {
+		m[name] = false
+	}
```
I think this is fine for explicit-ness, but not necessary. Just wondering, did you run into trouble with this or is this change just for "readability"?

jrajahalme replied:
Right, was not absolutely necessary, but thought keeping three places where the CRDs were listed "in sync" is unnecessary when can have it done only at one location.

jrajahalme replied:
So more for "maintainability" :-)

christarazi commented on pkg/k8s/synced/resources.go:
```diff
@@ -0,0 +1,128 @@
+	"k8s.io/client-go/tools/cache"
+)
+
+type Resources struct {
```
Nit: godoc

jrajahalme replied:
Added.

christarazi commented on pkg/k8s/synced/crd.go:
```diff
@@ -244,6 +244,8 @@ var (
 		crdResourceName(v2.CNName),
 		crdResourceName(v2.CIDName),
 		crdResourceName(v2.CLRPName),
+		crdResourceName(v2.CCLRPName),
```
Why is CCLRP also being added?

christarazi replied:
I believe it is also being removed in a pending PR: https://github.com/cilium/cilium/pull/13671/commits/33e8399a806fe217606721ca87118819ef16a20b

jrajahalme replied:
Only saw it was missing, was not aware of the PR for removing it. Will remove :-)

christarazi commented on clustermesh-apiserver/main.go:
```diff
@@ -543,6 +547,7 @@ func runServer(cmd *cobra.Command) {
 		if err := k8s.Init(k8sconfig.NewDefaultConfiguration()); err != nil {
 			log.WithError(err).Fatal("Unable to connect to Kubernetes apiserver")
 		}
+		synced.WaitForCRDsToRegister(context.TODO(), &synced.Resources{}, &synced.APIGroups{})
```
Is the `CRDWaitTimeout` supposed to be used to create a context and passed in here?

christarazi replied:
I'm also not entirely clear why we need to also declare the `CRDWaitTimeout` flag again in this file.

jrajahalme replied:
Seems not needed, as there is a fatal log upon CRDWaitTimeout within the waiter code. Bumped into it before I added the code to use the `CRDWaitTimeout` option.

christarazi commented on pkg/k8s/synced/crd.go:
```diff
@@ -12,7 +12,7 @@
 // limitations under the License.
 
-package watchers
+package synced
```
I think the refactor is good. I think a package name like `resources` would be more apt to represent the code inside it. What do you think?

jrajahalme replied:
My reading of "resources" has a connotation that it would include definition or more of the resource management... This is just a simple tracker for the initial resource sync. Started with `sync` as the package name, but soon figured that was pretty bad :-)

christarazi replied:
Fair enough.

> This is just a simple tracker for the initial resource sync

I think this is a good start to a package-level godoc. Could you create a `doc.go` and include a godoc including the above quoted text? I agree, naming this is kinda hard, so having a godoc to explain it will be valuable. Something like:

```
Package synced provides a simple tracker to ensure that Cilium K8s resources such as CRDs are synchronized in the cluster. It also ... blah blah (if there's a need to expand more...)
```

jrajahalme replied:
I'll leave this as-is for now, but don't oppose potential future changes for a better name.

christarazi reviewed (changes requested):
Overall looks good. Some items to iterate on.

christarazi commented on pkg/k8s/synced/crd.go:
```diff
@@ -41,21 +41,16 @@ import (
+	k8sAPIGroupCRD = "CustomResourceDefinition"
+)
+
+func WaitForCRDsToRegister(ctx context.Context, rs *Resources, ag *APIGroups) error {
```
Maybe we can consider renaming this function to `SyncCRDs` and having the following godoc:

```
SyncCRDs will sync Cilium CRDs to ensure that they have all been installed inside the K8s cluster. These CRDs are added by the Cilium Operator. This function will block until it finds all the CRDs or if a timeout occurs.

jrajahalme replied:
Done, thanks!

christarazi commented on pkg/k8s/synced/crd.go:
```diff
@@ -231,25 +226,35 @@ type crdState struct {
+func GetCRDResourceNames() []string {
+	return crdResourceNames
+}
+
```
Nit: Let's move this up in the file since it's exported. Generally, exported functions should come first, and then unexported / private APIs can follow.

jrajahalme replied:
Done.

christarazi commented on pkg/k8s/synced/resources.go:
```diff
@@ -0,0 +1,128 @@
+}
+
+// WaitForCacheSync waits for k8s caches to be synchronized for the given
+// resource. Returns once all resourcesNames are synchronized with cilium-agent.
```
```
// WaitForCacheSync waits for all K8s resources represented by resourceNames to have their K8s caches synchronized. 
```

jrajahalme replied:
Changed.

christarazi commented on pkg/k8s/synced/logfields.go:
```diff
@@ -0,0 +1,31 @@
+// Copyright 2016-2017 Authors of Cilium
```
Nit: 2020

jrajahalme replied:
Docs fixed before the merge

christarazi reviewed (approved):
LGTM! Just a few more non-blocking items and they don't require rerunning CI, so let's wait until CI finishes.

- ~CRDWaitTimeout is still being declared in clustermesh main.go~
- Package level godoc on package `synced`

jrajahalme committed:
k8s: Doc fixes

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
